File: frontend/src/pointer.ts
/**
 * RFC 6901 JSON Pointers, matching `backend/scribe40k/pointer.py`.
 *
 * Pointers are the shared vocabulary between the two halves of the app: flags address
 * fields by pointer, the assignment tray targets pointers, and autosave sends pointers.
 * Every editable control on the sheet is identified by one.
 */

export function escapeToken(token: string): string {
  return token.replace(/~/g, "~0").replace(/\//g, "~1");
}

export function ptr(...tokens: (string | number)[]): string {
  return tokens.map((token) => `/${escapeToken(String(token))}`).join("");
}

function parse(pointer: string): string[] {
  if (pointer === "") return [];
  if (!pointer.startsWith("/")) throw new Error(`not a JSON Pointer: ${pointer}`);
  return pointer
    .slice(1)
    .split("/")
    .map((token) => token.replace(/~1/g, "/").replace(/~0/g, "~"));
}

export function resolve<T = unknown>(document: unknown, pointer: string): T | undefined {
  let current: any = document;
  for (const token of parse(pointer)) {
    if (current === null || current === undefined) return undefined;
    current = Array.isArray(current) ? current[Number(token)] : current[token];
  }
  return current as T;
}

/** Immutable set: returns a copy with `pointer` changed, sharing everything untouched. */
export function withValue<T>(document: T, pointer: string, value: unknown): T {
  const tokens = parse(pointer);
  if (tokens.length === 0) return value as T;

  const clone = (node: any, depth: number): any => {
    const token = tokens[depth];
    const last = depth === tokens.length - 1;

    if (Array.isArray(node)) {
      const copy = [...node];
      const index = token === "-" ? copy.length : Number(token);
      copy[index] = last ? value : clone(node[index], depth + 1);
      return copy;
    }

    const copy = { ...(node ?? {}) };
    copy[token] = last ? value : clone(copy[token], depth + 1);
    return copy;
  };

  return clone(document, 0);
}

File: backend/scribe40k/pointer.py
"""RFC 6901 JSON Pointers.

Flags, the assignment tray and the editor's autosave all address fields by pointer, so
this is the shared vocabulary between backend and frontend. Small enough to implement
directly rather than take a dependency for.
"""

from __future__ import annotations

from typing import Any


def parse(pointer: str) -> list[str]:
    """Split a pointer into unescaped tokens. ``""`` is the whole document."""
    if pointer == "":
        return []
    if not pointer.startswith("/"):
        raise ValueError(f"a JSON Pointer must start with '/': {pointer!r}")
    return [token.replace("~1", "/").replace("~0", "~") for token in pointer[1:].split("/")]


def escape(token: str) -> str:
    return token.replace("~", "~0").replace("/", "~1")


def join(*tokens: object) -> str:
    return "".join(f"/{escape(str(token))}" for token in tokens)


_MISSING = object()


def resolve(document: Any, pointer: str, default: Any = _MISSING) -> Any:
    """Read the value at ``pointer``.

    Raises ``KeyError`` when the path does not exist and no default is given.
    """
    current = document
    for token in parse(pointer):
        try:
            current = current[int(token)] if isinstance(current, list) else current[token]
        except (KeyError, IndexError, TypeError, ValueError):
            if default is _MISSING:
                raise KeyError(f"{pointer} does not resolve") from None
            return default
    return current


def exists(document: Any, pointer: str) -> bool:
    return resolve(document, pointer, default=_MISSING) is not _MISSING


def set_value(document: Any, pointer: str, value: Any) -> None:
    """Write ``value`` at ``pointer``, in place.

    Intermediate containers must already exist. This is deliberate: the character document
    is created complete by :func:`scribe40k.blank.blank_character`, so a pointer that does
    not resolve is a bug or a hallucinated field, not something to paper over by inventing
    structure.

    Appending to an array uses the RFC's ``-`` token.
    """
    tokens = parse(pointer)
    if not tokens:
        raise ValueError("cannot replace the whole document")

    current = document
    for token in tokens[:-1]:
        current = current[int(token)] if isinstance(current, list) else current[token]

    last = tokens[-1]
    if isinstance(current, list):
        if last == "-":
            current.append(value)
        else:
            current[int(last)] = value
    else:
        current[last] = value


def deep_merge(target: dict, source: dict) -> list[str]:
    """Merge ``source`` into ``target`` in place; return the pointers that were written.

    Used to fold each section job's fragment into the blank character document. The rules
    suit that job specifically:

    * ``None`` in the source never overwrites a value in the target. A model omitting a
      field it could not read must not erase a printed constant.
    * Dicts merge recursively; lists replace wholesale, because a list on this sheet is a
      complete answer (all the gear, all the specialisations), not a patch.
    """
    written: list[str] = []
    _merge(target, source, "", written)
    return written


def _merge(target: dict, source: dict, prefix: str, written: list[str]) -> None:
    for key, value in source.items():
        path = f"{prefix}/{escape(key)}"

        if value is None:
            continue

        if isinstance(value, dict) and isinstance(target.get(key), dict):
            _merge(target[key], value, path, written)
        else:
            target[key] = value
            written.append(path)
